File: rust/src/streaming/watch_together.rs
// SPDX-License-Identifier: MIT

//! Watch-Together service: synchronized playback across session members.

use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

use crate::constants::DEFAULT_TTL;
use crate::protocol::{MeshPacket, PacketType};
use crate::routing::sender::MeshSender;
use crate::routing::RoutingService;

// ── Wire types ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct WatchSyncPayload {
    pub session_id: Uuid,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub playback_speed: Option<f64>,
    pub sent_at_ms: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct WatchReactionPayload {
    pub session_id: Uuid,
    pub reaction: String,
}

// ── Session state ──────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct WatchSession {
    pub session_id: Uuid,
    pub content_id: Option<String>,
    pub members: HashSet<String>,
    /// Current playback position in ms (at time of last sync).
    pub position_ms: i64,
    /// When the last sync was recorded (unix ms).
    pub last_sync_at_ms: i64,
    /// Current playback speed (1.0 = normal).
    pub playback_speed: f64,
    pub playing: bool,
}

// ── Service ───────────────────────────────────────────────────────────────────

pub struct WatchTogetherService {
    sender: Arc<dyn MeshSender>,
    routing: Arc<RoutingService>,
    sessions: Mutex<HashMap<Uuid, WatchSession>>,
}

impl WatchTogetherService {
    pub fn new(sender: Arc<dyn MeshSender>, routing: Arc<RoutingService>) -> Self {
        Self {
            sender,
            routing,
            sessions: Mutex::new(HashMap::new()),
        }
    }

    /// Invite members to a new or existing watch session. The content_id
    /// is broadcast to all invitees.
    pub async fn invite_to_session(
        &self,
        session_id: Uuid,
        content_id: &str,
        member_uhids: &[String],
    ) -> Result<(), String> {
        if member_uhids.is_empty() {
            return Err("member_uhids is empty".into());
        }
        let now = unix_millis();
        {
            let mut sessions = self.sessions.lock().unwrap();
            let entry = sessions.entry(session_id).or_insert_with(|| WatchSession {
                session_id,
                content_id: Some(content_id.to_string()),
                members: HashSet::new(),
                position_ms: 0,
                last_sync_at_ms: now,
                playback_speed: 1.0,
                playing: false,
            });
            for m in member_uhids {
                entry.members.insert(m.clone());
            }
        }
        for uhid in member_uhids {
            let sync = WatchSyncPayload {
                session_id,
                kind: "join".into(),
                position_ms: Some(0),
                playback_speed: Some(1.0),
                sent_at_ms: now,
                content_id: Some(content_id.to_string()),
            };
            self.send_sync(uhid, &sync).await;
        }
        Ok(())
    }

    /// Broadcast a play command to all session members.
    pub async fn play(&self, session_id: Uuid, position_ms: i64) -> Result<(), String> {
        let now = unix_millis();
        let (speed, members) = self.update_session(session_id, |s| {
            s.playing = true;
            s.position_ms = position_ms;
            s.last_sync_at_ms = now;
            (s.playback_speed, s.members.clone())
        })?;
        let sync = WatchSyncPayload {
            session_id,
            kind: "play".into(),
            position_ms: Some(position_ms),
            playback_speed: Some(speed),
            sent_at_ms: now,
            content_id: None,
        };
        self.broadcast_sync(session_id, &members, &sync).await;
        Ok(())
    }

    /// Broadcast a pause command.
    pub async fn pause(&self, session_id: Uuid, position_ms: i64) -> Result<(), String> {
        let now = unix_millis();
        let (_, members) = self.update_session(session_id, |s| {
            s.playing = false;
            s.position_ms = position_ms;
            s.last_sync_at_ms = now;
            (s.playback_speed, s.members.clone())
        })?;
        let sync = WatchSyncPayload {
            session_id,
            kind: "pause".into(),
            position_ms: Some(position_ms),
            playback_speed: None,
            sent_at_ms: now,
            content_id: None,
        };
        self.broadcast_sync(session_id, &members, &sync).await;
        Ok(())
    }

    /// Broadcast a seek command.
    pub async fn seek(&self, session_id: Uuid, position_ms: i64) -> Result<(), String> {
        let now = unix_millis();
        let (_, members) = self.update_session(session_id, |s| {
            s.position_ms = position_ms;
            s.last_sync_at_ms = now;
            (s.playback_speed, s.members.clone())
        })?;
        let sync = WatchSyncPayload {
            session_id,
            kind: "seek".into(),
            position_ms: Some(position_ms),
            playback_speed: None,
            sent_at_ms: now,
            content_id: None,
        };
        self.broadcast_sync(session_id, &members, &sync).await;
        Ok(())
    }

    /// Broadcast a playback speed change.
    pub async fn set_speed(&self, session_id: Uuid, playback_speed: f64) -> Result<(), String> {
        let now = unix_millis();
        let (pos, members) = self.update_session(session_id, |s| {
            s.playback_speed = playback_speed;
            s.last_sync_at_ms = now;
            (s.position_ms, s.members.clone())
        })?;
        let sync = WatchSyncPayload {
            session_id,
            kind: "speed".into(),
            position_ms: Some(pos),
            playback_speed: Some(playback_speed),
            sent_at_ms: now,
            content_id: None,
        };
        self.broadcast_sync(session_id, &members, &sync).await;
        Ok(())
    }

    /// Send an emoji/text reaction to all session members.
    pub async fn send_reaction(
        &self,
        session_id: Uuid,
        reaction: &str,
    ) -> Result<(), String> {
        let members = {
            let sessions = self.sessions.lock().unwrap();
            sessions
                .get(&session_id)
                .ok_or_else(|| format!("unknown session_id {}", session_id))?
                .members
                .clone()
        };
        let body = serde_json::to_vec(&WatchReactionPayload {
            session_id,
            reaction: reaction.to_string(),
        })
        .unwrap_or_default();
        let local = self.sender.local_uhid();
        for uhid in &members {
            if *uhid == local {
                continue;
            }
            let mut pkt = MeshPacket::new(PacketType::WatchReaction, local.clone());
            pkt.destination_uhid = uhid.clone();
            pkt.ttl = DEFAULT_TTL;
            pkt.payload = body.clone();
            self.sender.send(&pkt, uhid).await;
        }
        Ok(())
    }

    /// Route an inbound packet to the right handler.
    pub async fn handle_packet(&self, packet: &MeshPacket) -> Result<(), String> {
        match packet.packet_type {
            PacketType::WatchSync => self.handle_sync(packet).await,
            PacketType::WatchReaction => Ok(()), // surfaced to app layer
            _ => Ok(()),
        }
    }

    // ── private ────────────────────────────────────────────────────────────────

    /// Send a WatchSync to a single peer.
    async fn send_sync(&self, to_uhid: &str, sync: &WatchSyncPayload) {
        let body = match serde_json::to_vec(sync) {
            Ok(b) => b,
            Err(_) => return,
        };
        let route = self.routing.find_route(to_uhid).await;
        let next_hop = route
            .map(|r| r.next_hop_uhid)
            .unwrap_or_else(|| to_uhid.to_string());

        let mut pkt = MeshPacket::new(PacketType::WatchSync, self.sender.local_uhid());
        pkt.destination_uhid = to_uhid.to_string();
        pkt.ttl = DEFAULT_TTL;
        pkt.payload = body;
        self.sender.send(&pkt, &next_hop).await;
    }

    /// Broadcast a WatchSync to all members except self.
    async fn broadcast_sync(
        &self,
        _session_id: Uuid,
        members: &HashSet<String>,
        sync: &WatchSyncPayload,
    ) {
        let local = self.sender.local_uhid();
        for uhid in members {
            if *uhid == local {
                continue;
            }
            self.send_sync(uhid, sync).await;
        }
    }

    async fn handle_sync(&self, packet: &MeshPacket) -> Result<(), String> {
        let sync: WatchSyncPayload = serde_json::from_slice(&packet.payload)
            .map_err(|e| format!("bad watch sync payload: {}", e))?;

        let now = unix_millis();
        // RTT compensation: advance position by elapsed time * speed.
        let elapsed_ms = now - sync.sent_at_ms;
        let speed = sync.playback_speed.unwrap_or(1.0);
        let compensated_position = sync.position_ms.map(|pos| {
            pos + ((elapsed_ms as f64) * speed) as i64
        });

        let mut sessions = self.sessions.lock().unwrap();
        let entry = sessions.entry(sync.session_id).or_insert_with(|| WatchSession {
            session_id: sync.session_id,
            content_id: sync.content_id.clone(),
            members: HashSet::new(),
            position_ms: compensated_position.unwrap_or(0),
            last_sync_at_ms: now,
            playback_speed: speed,
            playing: false,
        });

        match sync.kind.as_str() {
            "join" => {
                entry.members.insert(packet.source_uhid.clone());
                if let Some(cid) = sync.content_id {
                    entry.content_id = Some(cid);
                }
            }
            "leave" | "end" => {
                entry.members.remove(&packet.source_uhid);
            }
            "play" => {
                entry.playing = true;
                entry.playback_speed = speed;
                if let Some(pos) = compensated_position {
                    entry.position_ms = pos;
                }
                entry.last_sync_at_ms = now;
            }
            "pause" => {
                entry.playing = false;
                if let Some(pos) = sync.position_ms {
                    entry.position_ms = pos;
                }
                entry.last_sync_at_ms = now;
            }
            "seek" => {
                if let Some(pos) = compensated_position {
                    entry.position_ms = pos;
                }
                entry.last_sync_at_ms = now;
            }
            "speed" => {
                entry.playback_speed = speed;
                entry.last_sync_at_ms = now;
            }
            _ => {}
        }
        Ok(())
    }

    /// Helper: mutate a session under lock, returning a value.
    fn update_session<T, F>(&self, session_id: Uuid, f: F) -> Result<T, String>
    where
        F: FnOnce(&mut WatchSession) -> T,
    {
        let mut sessions = self.sessions.lock().unwrap();
        let entry = sessions
            .get_mut(&session_id)
            .ok_or_else(|| format!("unknown session_id {}", session_id))?;
        Ok(f(entry))
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

fn unix_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

File: go/streaming/watch_together.go
// SPDX-License-Identifier: MIT

package streaming

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/bhengubv/aether-protocol/go/constants"
	"github.com/bhengubv/aether-protocol/go/protocol"
	"github.com/bhengubv/aether-protocol/go/routing"
)

// ──────────────────────────────────────────────
// JSON wire types
// ──────────────────────────────────────────────

// WatchSyncPayload is the JSON body for WatchSync packets.
type WatchSyncPayload struct {
	SessionID     string  `json:"session_id"`
	Kind          string  `json:"kind"` // "play" | "pause" | "seek" | "speed" | "invite"
	PositionMs    int64   `json:"position_ms,omitempty"`
	PlaybackSpeed float64 `json:"playback_speed,omitempty"`
	SentAtMs      int64   `json:"sent_at_ms"`
	ContentID     string  `json:"content_id,omitempty"`
}

// WatchReactionPayload is the JSON body for WatchReaction packets.
type WatchReactionPayload struct {
	SessionID string `json:"session_id"`
	Reaction  string `json:"reaction"`
}

// ──────────────────────────────────────────────
// Session state
// ──────────────────────────────────────────────

// WatchSession holds runtime state for a co-watch session.
type WatchSession struct {
	SessionID     uuid.UUID
	ContentID     string
	Members       map[string]struct{}
	PlaybackSpeed float64
	mu            sync.Mutex
}

// ──────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────

// WatchTogetherService synchronises media playback across mesh peers.
type WatchTogetherService struct {
	sender    routing.MeshSender
	localUhid string
	sessions  sync.Map // uuid.UUID -> *WatchSession

	// Event callbacks — optional.
	OnInviteReceived func(session *WatchSession, fromUhid string)
	OnSyncReceived   func(session *WatchSession, fromUhid string, positionMs int64, speed float64)
	OnReaction       func(session *WatchSession, fromUhid, reaction string)
}

// NewWatchTogetherService constructs a WatchTogetherService.
func NewWatchTogetherService(sender routing.MeshSender) *WatchTogetherService {
	if sender == nil {
		panic("streaming: watch sender must not be nil")
	}
	return &WatchTogetherService{
		sender:    sender,
		localUhid: sender.LocalUhid(),
	}
}

// InviteToSession invites members to a co-watch session for contentID.
func (s *WatchTogetherService) InviteToSession(ctx context.Context, sessionID uuid.UUID, contentID string, memberUhids []string) error {
	if len(memberUhids) == 0 {
		return errors.New("streaming: memberUhids must not be empty")
	}
	session := s.getOrCreateWatchSession(sessionID, contentID)

	session.mu.Lock()
	for _, m := range memberUhids {
		session.Members[m] = struct{}{}
	}
	session.mu.Unlock()

	payload := WatchSyncPayload{
		SessionID: sessionID.String(),
		Kind:      "invite",
		ContentID: contentID,
		SentAtMs:  time.Now().UnixMilli(),
	}
	for _, m := range memberUhids {
		if err := s.sendWatchSync(ctx, m, payload); err != nil {
			return err
		}
	}
	return nil
}

// Play sends a play-sync to all session members.
func (s *WatchTogetherService) Play(ctx context.Context, sessionID uuid.UUID, positionMs int64) error {
	return s.sendSyncToAll(ctx, sessionID, "play", positionMs, 0)
}

// Pause sends a pause-sync to all session members.
func (s *WatchTogetherService) Pause(ctx context.Context, sessionID uuid.UUID, positionMs int64) error {
	return s.sendSyncToAll(ctx, sessionID, "pause", positionMs, 0)
}

// Seek sends a seek-sync to all session members.
func (s *WatchTogetherService) Seek(ctx context.Context, sessionID uuid.UUID, positionMs int64) error {
	return s.sendSyncToAll(ctx, sessionID, "seek", positionMs, 0)
}

// SetSpeed sends a speed-change sync to all session members.
func (s *WatchTogetherService) SetSpeed(ctx context.Context, sessionID uuid.UUID, playbackSpeed float64) error {
	session, err := s.getWatchSession(sessionID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	session.PlaybackSpeed = playbackSpeed
	session.mu.Unlock()
	return s.sendSyncToAll(ctx, sessionID, "speed", 0, playbackSpeed)
}

// SendReaction broadcasts a reaction emoji/string to all session members.
func (s *WatchTogetherService) SendReaction(ctx context.Context, sessionID uuid.UUID, reaction string) error {
	if reaction == "" {
		return errors.New("streaming: reaction must not be empty")
	}
	session, err := s.getWatchSession(sessionID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	members := s.watchMemberSnapshot(session)
	session.mu.Unlock()

	body, err := json.Marshal(WatchReactionPayload{
		SessionID: sessionID.String(),
		Reaction:  reaction,
	})
	if err != nil {
		return fmt.Errorf("streaming: marshal reaction: %w", err)
	}

	for _, m := range members {
		if m == s.localUhid {
			continue
		}
		pkt := protocol.NewMeshPacket()
		pkt.Type = protocol.WatchReaction
		pkt.SourceUhid = s.localUhid
		pkt.DestinationUhid = m
		pkt.Ttl = constants.DefaultTtl
		pkt.Payload = body
		_, _ = s.sender.Send(ctx, pkt, m)
	}
	return nil
}

// HandlePacket processes inbound WatchSync or WatchReaction packets.
func (s *WatchTogetherService) HandlePacket(_ context.Context, packet *protocol.MeshPacket) error {
	if packet == nil {
		return errors.New("streaming: watch packet must not be nil")
	}
	switch packet.Type {
	case protocol.WatchSync:
		return s.handleWatchSync(packet)
	case protocol.WatchReaction:
		return s.handleWatchReaction(packet)
	default:
		return fmt.Errorf("streaming: unexpected watch packet type %s", packet.Type)
	}
}

// ──────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────

func (s *WatchTogetherService) handleWatchSync(packet *protocol.MeshPacket) error {
	var payload WatchSyncPayload
	if err := json.Unmarshal(packet.Payload, &payload); err != nil {
		return fmt.Errorf("streaming: unmarshal watch sync: %w", err)
	}
	sessionID, err := uuid.Parse(payload.SessionID)
	if err != nil {
		return fmt.Errorf("streaming: invalid session_id: %w", err)
	}

	switch payload.Kind {
	case "invite":
		session := s.getOrCreateWatchSession(sessionID, payload.ContentID)
		session.mu.Lock()
		session.Members[packet.SourceUhid] = struct{}{}
		session.mu.Unlock()
		if cb := s.OnInviteReceived; cb != nil {
			cb(session, packet.SourceUhid)
		}

	case "play", "pause", "seek", "speed":
		session, err := s.getWatchSession(sessionID)
		if err != nil {
			return nil // unknown session — drop
		}
		// RTT compensation: adjust position for network latency.
		// compensated = positionMs + elapsed_since_send * playbackSpeed
		speed := payload.PlaybackSpeed
		if speed == 0 {
			session.mu.Lock()
			speed = session.PlaybackSpeed
			if speed == 0 {
				speed = 1.0
			}
			session.mu.Unlock()
		}
		elapsedMs := time.Now().UnixMilli() - payload.SentAtMs
		compensatedPos := payload.PositionMs + int64(float64(elapsedMs)*speed)

		if cb := s.OnSyncReceived; cb != nil {
			cb(session, packet.SourceUhid, compensatedPos, speed)
		}
	}
	return nil
}

func (s *WatchTogetherService) handleWatchReaction(packet *protocol.MeshPacket) error {
	var payload WatchReactionPayload
	if err := json.Unmarshal(packet.Payload, &payload); err != nil {
		return fmt.Errorf("streaming: unmarshal watch reaction: %w", err)
	}
	sessionID, err := uuid.Parse(payload.SessionID)
	if err != nil {
		return fmt.Errorf("streaming: invalid session_id: %w", err)
	}
	session, err := s.getWatchSession(sessionID)
	if err != nil {
		return nil // unknown session — drop
	}
	if cb := s.OnReaction; cb != nil {
		cb(session, packet.SourceUhid, payload.Reaction)
	}
	return nil
}

func (s *WatchTogetherService) sendSyncToAll(ctx context.Context, sessionID uuid.UUID, kind string, positionMs int64, speed float64) error {
	session, err := s.getWatchSession(sessionID)
	if err != nil {
		return err
	}
	session.mu.Lock()
	members := s.watchMemberSnapshot(session)
	session.mu.Unlock()

	payload := WatchSyncPayload{
		SessionID:     sessionID.String(),
		Kind:          kind,
		PositionMs:    positionMs,
		PlaybackSpeed: speed,
		SentAtMs:      time.Now().UnixMilli(),
	}
	for _, m := range members {
		if m == s.localUhid {
			continue
		}
		if err := s.sendWatchSync(ctx, m, payload); err != nil {
			return err
		}
	}
	return nil
}

func (s *WatchTogetherService) sendWatchSync(ctx context.Context, toUhid string, payload WatchSyncPayload) error {
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("streaming: marshal watch sync: %w", err)
	}
	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.WatchSync
	pkt.SourceUhid = s.localUhid
	pkt.DestinationUhid = toUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body
	_, err = s.sender.Send(ctx, pkt, toUhid)
	return err
}

func (s *WatchTogetherService) getWatchSession(sessionID uuid.UUID) (*WatchSession, error) {
	v, ok := s.sessions.Load(sessionID)
	if !ok {
		return nil, fmt.Errorf("streaming: unknown watch session %s", sessionID)
	}
	return v.(*WatchSession), nil
}

func (s *WatchTogetherService) getOrCreateWatchSession(sessionID uuid.UUID, contentID string) *WatchSession {
	v, _ := s.sessions.LoadOrStore(sessionID, &WatchSession{
		SessionID:     sessionID,
		ContentID:     contentID,
		Members:       make(map[string]struct{}),
		PlaybackSpeed: 1.0,
	})
	return v.(*WatchSession)
}

func (s *WatchTogetherService) watchMemberSnapshot(session *WatchSession) []string {
	out := make([]string, 0, len(session.Members))
	for m := range session.Members {
		out = append(out, m)
	}
	return out
}
